## CytoTRACE.R takes in QC-processed RNA-seq data and calculates cytoTRACE values for all cells
## Requires : cytoTRACE, Matrix
## Inputs: _matrix.mtx, _barcodes.tsv, _genes.tsv, _phenotypes.tsv

### Command inputs
arg = commandArgs(trailingOnly=TRUE)
#arg[1] = expression matrix
#arg[2] = output name

###load cytotrace
# contains:
# CytoTRACE(): function to run CytoTRACE on a custom scRNA-seq dataset
# iCytoTRACE: function to run CytoTrACE across multiple, heterogeneous scRNA-seq batches/dataset
# plotCytoTRACE: function to generate 2D visualizations of CytoTRACE, phenotypes, and gene expression

library(CytoTRACE)

### load RNAseq data
# load libraries
library(Matrix)
# load expression matrix
expression_matrix <- readMM(paste0(arg[1],"_matrix.mtx"))
## Add cell names 
colnames(expression_matrix) <- read.delim(paste0(arg[1],"_barcodes.tsv"),header=FALSE)$V1
## Add gene names
rownames(expression_matrix) <- read.delim(paste0(arg[1],"_genes.tsv"),header=FALSE)$V2
# convert from dgTMatrix to dgCMatrix
expression_matrix <- as.data.frame(as.matrix(expression_matrix)) #convert from dgTMatrix to data.frame


#Create phenotype info
phenotypes <- read.delim(paste0(arg[1],"_phenotypes.tsv"),header=TRUE)
pheno_string <- phenotypes$Phenotype #phenotype string
names(pheno_string) <- rownames(phenotypes) #add barcode label names 


### run CytoTrace
results <- CytoTRACE(expression_matrix)


### Plot results
#     plotCytoTRACE(results, options...)
#       emb = ... to use your own coordinates generated by another algorithm
#       phenotype = ...to generate a 2D plot with phenotype mapping
#       gene = ... to generate a 2D plot with gene expression mapping
#     plotCytoGenes(results, options...)
#       numofGenes = ... number of genes to display

plotCytoTRACE(results,phenotype=pheno_string, outputDir=arg[2])


plotCytoGenes(results, numOfGenes = 10, outputDir=arg[2])


