## CytoTRACE.R takes in QC-processed RNA-seq data and calculates cytoTRACE values for all cells
## Requires : Matrix, CytoTraceMod.R

### Command inputs
arg = commandArgs(trailingOnly=TRUE)
#arg[1] = expression matrix
#arg[2] = phenotype vector 
#arg[3] = output name

###load cytotrace
# contains:
# CytoTRACE(): function to run CytoTRACE on a custom scRNA-seq dataset
# iCytoTRACE: function to run CytoTrACE across multiple, heterogeneous scRNA-seq batches/dataset
# plotCytoTRACE: function to generate 2D visualizations of CytoTRACE, phenotypes, and gene expression

library(CytoTRACE)

###load cytotrace function modified for dgCMatrix format
source("CytoTraceMod.R")

### load RNAseq data
# load libraries
library(Matrix)
# load expression matrix
expression_matrix <- readMM(arg[1])
# convert from dgTMatrix to dgCMatrix
expression_matrix <- as(expression_matrix[1:dim(expression_matrix)[1],],"dgCMatrix") #subset matrix and then convert from dgTMatrix to dgCMatrix


#Create phenotype info
pheno_string <- read.delim(arg[2],',',header=FALSE)$V2 #phenotype string
names(pheno_string) <- variable.names(expression_matrix) #add barcode label names 


### run CytoTrace
results <- CytoTRACE_mod(expression_matrix)


### Plot results
#     plotCytoTRACE(results, options...)
#       emb = ... to use your own coordinates generated by another algorithm
#       phenotype = ...to generate a 2D plot with phenotype mapping
#       gene = ... to generate a 2D plot with gene expression mapping
#     plotCytoGenes(results, options...)
#       numofGenes = ... number of genes to display

plotCytoTRACE(results,phenotype=pheno_string, outputDir=paste0("./CytoTrace Results/",arg[3],"_"))


plotCytoGenes(results, numOfGenes = 10, outputDir=paste0("./CytoTrace Results/",arg[3],"_"))